tests/test_report.py

```diff
@@ -0,0 +1,324 @@
+        mock_post.return_value.status_code = 200
+        mock_post.return_value.json = access_token
+        report.handler({"emails": ["aa@gmail.com"]})
+        headers = {'Authorization': f'Bearer {access_token()["access_token"]}'}
+        check_mock_urlopen(mock_urlopen, base_headers=headers)
+
+
+@patch('urllib.request.urlopen')
+def test_http_server_error(mock_urlopen, kms):
+    """Test if HTTP error without JSON payload is handled"""
+
+    with patch('reporting.report.requests.post') as mock_post:
+        set_up(kms)
+        setup_slack_api(slack_api_response_ok)
+
+        err = urllib.error.HTTPError(
+            'http://testserver.com',
+            500,
+            'Some Error',
+            {'Xpto': 'abc'},
+            io.BytesIO(b'xpto'),
+        )
+        configure_mock_urlopen(mock_urlopen, err)
+        mock_post.return_value.status_code = 200
+        mock_post.return_value.json = access_token
+        response = report.handler({"emails": ["aa@gmail.com"]})
+        assert response['statusCode'] == 500
+
+
+@patch('urllib.request.urlopen')
+def test_slack_auth_token_is_not_valid(mock_urlopen, kms):
+    """Test if slack token is invalid"""
+
+    with patch('reporting.report.requests.post') as mock_post:
+        set_up(kms)
+        setup_slack_api(slack_api_response_invalid)
+        configure_mock_urlopen(mock_urlopen, graphql_response_ok)
+        mock_post.return_value.status_code = 200
+        mock_post.return_value.json = access_token
+        response = report.handler({"emails": ["aa@gmail.com"]})
+        assert response['statusCode'] == 401
+
+
+@patch('urllib.request.urlopen')
+def test_multiple_emails_are_passed(mock_urlopen, kms):
+    """Test if multiple emails are passed"""
+
+    with patch('reporting.report.requests.post') as mock_post:
+        set_up(kms)
+        setup_slack_api(slack_api_response_ok)
+        configure_mock_urlopen(mock_urlopen, graphql_response_ok)
+        mock_post.return_value.status_code = 200
+        mock_post.return_value.json = access_token
+        report.handler({"emails": ["aa@gmail.com", "bb@gmail.com"]})
+        headers = {'Authorization': f'Bearer {access_token()["access_token"]}'}
+        check_mock_urlopen(mock_urlopen, base_headers=headers)
+
+
+@patch('urllib.request.urlopen')
+def test_when_no_emails_are_passed(mock_urlopen, kms):
+    """Test when no emails are passed then it should not slack the report"""
+
+    with patch('reporting.report.requests.post') as mock_post:
+        set_up(kms)
+        configure_mock_urlopen(mock_urlopen, graphql_response_ok)
+        mock_post.return_value.status_code = 200
+        mock_post.return_value.json = access_token
+        report.handler()
+        df = pandas.read_csv(csv_file_path)
+        assert len(df) == 1
+
+
+@patch('urllib.request.urlopen')
+def test_when_empty_email_list_are_passed(mock_urlopen, kms):
+    """Test when empty emails list are passed then it should not slack the report"""
```

Comment:
Maybe reword slightly: `Test no slack message sent when empty email address list is provdided`